docs/models/advanced/extended-blockstates.md

```diff
@@ -0,0 +1,61 @@
+Extended Blockstates
+====================
+
+Extended blockstates provide a way for blocks to pass arbitrary data to their models. Ordinary blockstates and properties occupy only a fixed set of possible states, while extended states can form infinite sets. This accomplished through the use of unlisted properties (`IUnlistedProperty<V>`), extended block states (`IExtendedBlockState`), and extended block state containers (`ExtendedBlockState`).
+
+Ordinary block state containers (`BlockStateContainer`) take a set of listed properties (`IProperty<V>`) that define all possible values for themselves and use those to create all possible combinatons of values. These combinations are then turned into `IBlockState`s and stored in the finite set of all possible states. The properties are called "listed" as they appear on the F3 debug screen, all the way to the right.
+
+Extended block state containers take a set of listed properties, and also a set of *unlisted* properties. Unlisted properties' values can be anything that matches their type, and are not limited to a finite set. However, values are still required to satisfy the predicate `IUnlistedProperty::isValid`. The listed properties are again used to create the set of states, this time `IExtendedBlockState`s, but the unlisted properties remain unset. The unlisted properties are only set when `IExtendedBlockState::withProperty` is called to do so.
+
+One possible use case for this is to have advanced camouflaging blocks. The camouflaging block may have a tile entity that holds the `IBlockState` representative of the camouflagee, and an unlisted property to hold that state during rendering. This property can then be filled by `getExtendedState`. Finally, a custom [`IBakedModel`][IBakedModel] can steal the model for that state and use it instead of using the uncamouflaged model.
+
+Most of the methods on `IExtendedBlockState` are self-explanatory, maybe with the exception of `getClean`. `getClean` returns the base `IBlockState` with none of the unlisted properties set.
+
+Why Extended Blockstates
+------------------------
+
+Why use extended blockstates at all? Why not simply pass an `IBlockAccess` and `BlockPos` into the rendering system directly and have the [`IBakedModel`][IBakedModel] itself deal with it? Indeed, extended blockstates allow this to happen anyway! Why have the system at all when we could just do that? The reason is twofold.
+
+First, rendering is threaded, meaning that the world continues to chug along on the main thread while models are being rendered. However, in Minecraft, `World`s are simply not made for concurrent access. This makes the above solution downright *dangerous*, as it would entail multiple threads concurrently accessing the world. Therefore, extended states are used to collect all the data required for the model safely into the `IExtendedBlockState`. These objects can now be passed on to different threads.
```

Comment:
`getExtendedState` is called on the same thread, directly before getQuads, so this is a pretty weak justification. There is still the same threading concerns inside extended state methods.